serializer-base64/src/main/java/com/vlkan/hrrs/serializer/base64/Base64HttpRequestRecordWriter.java

```diff
@@ -5,45 +5,58 @@
 
 public class Base64HttpRequestRecordWriter implements HttpRequestRecordWriter<String> {
 
-    private final Base64Encoder encoder;
+    protected final Base64Encoder encoder;
+    protected final HttpRequestRecordWriterTarget<String> target;
+    private final DateFormat dateFormat = new SimpleDateFormat("yyyyMMdd-HHmmss.SSSZ");
 
-    private final HttpRequestRecordWriterTarget<String> target;
-
-    public Base64HttpRequestRecordWriter(HttpRequestRecordWriterTarget<String> target, Base64Encoder encoder) {
+    protected Base64HttpRequestRecordWriter(HttpRequestRecordWriterTarget<String> target, Base64Encoder encoder) {
         this.target = checkNotNull(target, "target");
         this.encoder = checkNotNull(encoder, "encoder");
     }
 
+    public static Base64HttpRequestRecordWriter createBase64HttpRequestRecordWriter(HttpRequestRecordWriterTarget<String> target, Base64Encoder encoder) {
+        if (System.getProperty("use.hrrs.queue") != null) {
+            return new QueueBase64HttpRequestRecordWriter(target, encoder);
+        }
+        return new Base64HttpRequestRecordWriter(target, encoder);
+    }
+
     @Override
     public HttpRequestRecordWriterTarget<String> getTarget() {
         return target;
     }
 
     @Override
-    public void write(HttpRequestRecord record) {
+    public void write(HttpRequestRecord record) throws IOException {
         try {
+            String formattedDate = dateFormat.format(record.getTimestamp());
+            StringBuilder toBeRecorded = new StringBuilder(512);
+            byte[] recordBytes = writeRecord(record);
+            String encodedRecordBytes = encoder.encode(recordBytes);
+            toBeRecorded
+                    .append(record.getId())
+                    .append(FIELD_SEPARATOR)
+                    .append(formattedDate)
+                    .append(FIELD_SEPARATOR)
+                    .append(record.getGroupName())
+                    .append(FIELD_SEPARATOR)
+                    .append(record.getMethod().toString())
+                    .append(FIELD_SEPARATOR)
+                    .append(encodedRecordBytes)
+                    .append(RECORD_SEPARATOR);
             synchronized (this) {
-                target.write(record.getId());
-                target.write(FIELD_SEPARATOR);
-                target.write(DATE_FORMAT.format(record.getTimestamp()));
-                target.write(FIELD_SEPARATOR);
-                target.write(record.getGroupName());
-                target.write(FIELD_SEPARATOR);
-                target.write(record.getMethod().toString());
-                target.write(FIELD_SEPARATOR);
-                byte[] recordBytes = writeRecord(record);
-                String encodedRecordBytes = encoder.encode(recordBytes);
-                target.write(encodedRecordBytes);
-                target.write(RECORD_SEPARATOR);
+                target.write(toBeRecorded.toString());
```

Comment:
Indeed this commit shortens the _LoC_ in the `synchronized` block, though I am not convinced that this performs better. You replace the lengthy `synchronized` block with a one-liner. Though the amount of bytes need be written to the `target` are exactly the same, this new scheme only avoids the time spent in `encode(byte[])`. Further, one also need to pay the cost of extra `StringBuilder` allocation each time. This allocation cost can be avoided using an object pool, but that is another story. In conclusion, do you have empirical evidence to prove that this approach performs better? Did it cause a bottleneck in your case?